### ISSUE TITLE
Close Date Display

### ISSUE THREAD

**Issue by ClaireValdivia** [usdigitalresponse/usdr-gost#2571]
### Why is this issue important?

sometimes close date does not show a date
Design: https://www.figma.com/file/yQx55VTr5awPYClHIJCz8y/Grant-Details?type=design&node-id=1480-3099&mode=dev

*Claire to add in file with examples of close date messages

### Current State

na

### Expected State

na

### Implementation Plan

The following functions need to be added...

**Comment by ClaireValdivia** [comment]
what was pulled before was scraped from grants.gov and not our xml so fields aren't exactly matching - Jeff needing to do some additional research to assess xml

**Comment by jeffsmohan** [comment]
@ClaireValdivia I've run an analysis of what close date explanation is available in the current dataset we have access to (based on a staging database dump from December). It looks like we have access to a field that gives us what we need, and very closely matches the original scraping-based analysis linked in #2066. See the analysis here for details: 

- Walkthrough: https://docs.google.com/document/d/17OvDqGCtw5YqHKHT7xoqkPwqvxJ5HK21B4QrS2VzrCw/edit
- Data sheet: https://docs.google.com/spreadsheets/d/10VB-THr0Qn1tmmAnWGkhE8_GMvtQ6I_i8yvLgpqHhBA/edit#gid=1506004290

@TylerHendrickson Assuming the above analysis checks out, I believe that means we need add a column to the database for `close_date_explanation` and update the grants ingest pipeline to populate it. Adding something like `grant.close_date_explanation = milestones?.close?.explanation ?? undefined;` to the `mapSourceDataToGrant` method is what I'm thinking. But I'd need some guidance on how to introduce new fields to the pipeline and database. 

(Alternatively, `raw_body` json is already loaded into the db, so we could simply read that out of the json on the server or even on the client. But that doesn't seem as clean/maintainable as the "correct" solution implemented in the ingest pipeline. Lmk what you think though.)

**Comment by TylerHendrickson** [comment]
@jeffsmohan Your analysis looks good to me!

To enumerate next steps:
1. Create a Knex migration that adds a `close_date_explanation` (default null, type text) column to the `grants` table in the database.
2. As you noted, update the [`mapSourceDataToGrant()` function in `packages/server/src/lib/grants-ingest.js`](https://github.com/usdigitalresponse/usdr-gost/blob/74b150bc6d77aca4bf188a65261e5603b08a4b51/packages/server/src/lib/grants-ingest.js#L24-L65) to populate the new column from `source.opportunity.milestones.close.explanation`. 
    - Note that the `source` argument to `mapSourceDataToGrant()` is an object that conforms to the [`Grant` schema in the `grants-ingest` OpenAPI spec](https://github.com/usdigitalresponse/grants-ingest/blob/986b36b12f96c6e4b37302b0617678d2ae6c2ac7/openapi/openapi.yaml#L268-L317).
3. Add an additional Knex migration that backfills the new `grants.close_date_explanation` column using the `grants.raw_body_json` column where `close_date_explanation` is null (this is necessary to ensure that any data ingested from the pipeline between Steps 1 and 3 is not overwritten by this backfill operation). 
    - As a point of reference, [this migration is doing something very similar](https://github.com/usdigitalresponse/usdr-gost/blob/main/packages/server/migrations/20240119214120_populate_funding_activity_category_codes_column.js) (in that case, the migration is backfilling `grants.funding_activity_category_codes` from `grants.raw_body_json`).

NB: I don't believe that any changes to the [`usdigitalresponse/grants-ingest` pipeline service](https://github.com/usdigitalresponse/grants-ingest/) should be necessary – it's already providing the data to gost, so it's just a matter of ensuring that going forward, the new column is populated whenever gost receives new updates from the pipeline.

**Comment by ClaireValdivia** [comment]
@jeffsmohan thanks for your work on this! just want to confirm that there is nothing to test in staging right now based on this first PR, is that correct?

**Comment by jeffsmohan** [comment]
@ClaireValdivia Of course! There should be no user-facing changes on staging this week. The feature needs to be phased across multiple deployments. I believe @TylerHendrickson helped get the first PR (https://github.com/usdigitalresponse/usdr-gost/pull/2648) into staging for today's deploy. He said he was going to verify the database migration that introduces. 

But tldr, no user-facing changes on staging currently.

**Comment by ClaireValdivia** [comment]
This is looking great. confirmed that "current closing date" shows when a grant has different data for original vs current closing date.
